src/lib/operators/case_operator.cpp

```diff
@@ -0,0 +1,226 @@
+     * materialized Data Table, we re-use the columns from the input chunk if they are Data Columns and materialize them
+     * if they are Reference Columns.
+     */
+    ChunkColumns output_columns;
+    if (input_table_left()->type() == TableType::References) {
+      for (const auto& column : input_chunk->columns()) {
+        output_columns.emplace_back(materialize_as_value_column(*column));
+      }
+    } else {
+      output_columns = input_chunk->columns();
+    }
+
+    /**
+     * Create an output_column from each CaseExpression
+     */
+
+    // If a column in the input chunk is used as a WHEN column, it is materialized into this cache
+    std::vector<std::optional<std::vector<int32_t>>> materialized_when_columns(input_table_left()->column_count());
+
+    for (const auto& abstract_case_expression : _case_expressions) {
+      resolve_data_type(abstract_case_expression->result_data_type, [&](auto type) {
+        using ResultDataType = typename decltype(type)::type;
+        using NullOrValue = std::pair<bool, ResultDataType>;
+
+        // If a column in the input chunk is used as a THEN column, it is materialized into this cache
+        std::vector<std::optional<std::vector<NullOrValue>>> materialized_then_columns(
+            input_table_left()->column_count());
+
+        const auto& case_expression =
+            static_cast<const PhysicalCaseExpression<ResultDataType>&>(*abstract_case_expression);
+
+        pmr_concurrent_vector<ResultDataType> output_values(input_chunk->size());
+        pmr_concurrent_vector<bool> output_nulls(input_chunk->size());
+
+        /**
+         * Compute the CaseExpression for each row in the Chunk.
+         */
+        for (auto chunk_offset = ChunkOffset{0}; chunk_offset < input_chunk->size(); ++chunk_offset) {
+          auto any_when_claused_fulfilled = false;  // If no WHEN clause was TRUE, we have to use the ELSE clause
+
+          /**
+           * Find the first case_clause whose WHEN is true and write its THEN to output_values/nulls
+           */
+          for (const auto& case_clause : case_expression.clauses) {
+            // If not yet done, materialize the WHEN column and cache it.
+            auto& materialized_when_column = materialized_when_columns[case_clause.when];
+            if (!materialized_when_column) {
+              materialized_when_column.emplace();
+              materialized_when_column->reserve(input_chunk->size());
+              materialize_values(*input_chunk->get_column(case_clause.when), *materialized_when_column);
+            }
+
+            // The case_clause is not true? Continue with the next clause.
+            if ((*materialized_when_column)[chunk_offset] == int32_t{0}) continue;
+
+            const auto null_and_value =
+                materialize_case_result(case_clause.then, *input_chunk, chunk_offset, materialized_then_columns);
```

Comment:
I wonder if it would be better to check the cases in a different order: 1. check first case for all chunk_offsets and copy results 2. check second … 

The caches help probably but it’s still a lot of jumping around.

Reply:
Sequential reads are better, yes, but having a handful of sequentially increased pointers should have the same advantage.

Reply:
The it is written now, multiple columns might be read interleaved, which is not very cache friendly. Also there is all this overhead of checking the result type (null, column, value), which could be done once for all. If we checked the cases first for all records, we wouldn’t have to materialise entire columns either. There a lots of advantages.